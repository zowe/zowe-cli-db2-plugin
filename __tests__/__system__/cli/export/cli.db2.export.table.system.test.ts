/*
* This program and the accompanying materials are made available under the terms of the *
* Eclipse Public License v2.0 which accompanies this distribution, and is available at *
* https://www.eclipse.org/legal/epl-v20.html                                      *
*                                                                                 *
* SPDX-License-Identifier: EPL-2.0                                                *
*                                                                                 *
* Copyright Contributors to the Zowe Project.                                     *
*                                                                                 *
*/

import { ITestEnvironment } from "../../../__src__/environment/doc/response/ITestEnvironment";
import { TestEnvironment } from "../../../__src__/environment/TestEnvironment";
import { runCliScript } from "../../../__src__/TestUtils";
import * as fs from "fs";

let TEST_ENV: ITestEnvironment;
let host: string;
let port: number;
let user: string;
let password: string;
let database: string;

describe("db2 export table command", () => {

    // Create the unique test environment
    beforeAll(async () => {
        TEST_ENV = await TestEnvironment.setUp({
            installPlugin: true,
            tempProfileTypes: ["db2"],
            testName: "export_table_command",
        });

        host = TEST_ENV.systemTestProperties.db2.host;
        port = TEST_ENV.systemTestProperties.db2.port;
        user = TEST_ENV.systemTestProperties.db2.user;
        password = TEST_ENV.systemTestProperties.db2.password;
        database = TEST_ENV.systemTestProperties.db2.database;
    });

    afterAll(async () => {
        await TestEnvironment.cleanUp(TEST_ENV);
    });

    it("should display the help", () => {
        const response = runCliScript(__dirname + "/__scripts__/table_help.sh", TEST_ENV);
        expect(response.stdout.toString()).toMatchSnapshot();
        expect(response.stderr.toString()).toBe("");
        expect(response.status).toBe(0);
    });

    it("should fail with invalid option", async () => {
        const response = runCliScript(__dirname + "/__scripts__/table_invalid_option.sh", TEST_ENV);
        expect(response.stderr.toString()).toMatchSnapshot();
        expect(response.stdout.toString()).toBe("");
        expect(response.status).toBe(1);
    });

    it("should fail with empty table name", async () => {
        const response = runCliScript(__dirname + "/__scripts__/fail_empty_table.sh", TEST_ENV);
        expect(response.stderr.toString()).toMatchSnapshot();
        expect(response.stdout.toString()).toBe("");
        expect(response.status).toBe(1);
    });

    it("should fail to export a not existing table", () => {
        const response = runCliScript(__dirname + "/__scripts__/fail_table_not_exists.sh", TEST_ENV);
        expect(response.stderr.toString()).toContain("Error getting metadata for the table");
        expect(response.stdout.toString()).toBe("");
        expect(response.status).toBe(1);
    });

    it("should export the system roles table to the standard out", () => {
        const regex: RegExp = new RegExp(fs.readFileSync(__dirname + "/__regex__/export_table.regex").toString(), "gs");
        const response = runCliScript(__dirname + "/__scripts__/success_export_stdout.sh", TEST_ENV);
        expect(response.stderr.toString()).toBe("");
        expect(response.status).toBe(0);
        expect(new RegExp(regex, "gs").test(response.stdout.toString())).toBe(true);
    });

    it("should export the system roles table to a file", () => {
        const regex = fs.readFileSync(__dirname + "/__regex__/export_table.regex").toString();
        const response = runCliScript(__dirname + "/__scripts__/success_export_file.sh", TEST_ENV);
        const outFilePath = TEST_ENV.workingDir + "/export.sql";
        expect(response.stdout.toString()).toBe("");
        expect(response.stderr.toString()).toBe("");
        expect(response.status).toBe(0);
        expect(fs.existsSync(outFilePath)).toBe(true);
        const outFile = fs.readFileSync(outFilePath).toString();
        fs.unlinkSync(outFilePath);
        expect(new RegExp(regex, "gs").test(outFile)).toBe(true);
    });

    it("should export the system roles table with separator", () => {
        const regex = fs.readFileSync(__dirname + "/__regex__/export_table.regex").toString();
        const response = runCliScript(__dirname + "/__scripts__/success_export_separator.sh", TEST_ENV);
        expect(response.stderr.toString()).toBe("");
        expect(response.status).toBe(0);
        expect(new RegExp(regex, "gs").test(response.stdout.toString())).toBe(true);
    });

    it("should export the system roles table to the standard out overriding some profile options by cmd arguments", () => {
        const regex: RegExp = new RegExp(fs.readFileSync(__dirname + "/__regex__/export_table.regex").toString(), "gs");
        const response = runCliScript(__dirname + "/__scripts__/success_override_profile.sh", TEST_ENV,
            [host, port, database]);
        expect(response.stderr.toString()).toBe("");
        expect(response.status).toBe(0);
        expect(new RegExp(regex, "gs").test(response.stdout.toString())).toBe(true);
    });

});

describe("db2 export table command without profile", () => {

    // Create the unique test environment
    beforeAll(async () => {
        TEST_ENV = await TestEnvironment.setUp({
            installPlugin: true,
            testName: "export_table_command",
        });

        host = TEST_ENV.systemTestProperties.db2.host;
        port = TEST_ENV.systemTestProperties.db2.port;
        user = TEST_ENV.systemTestProperties.db2.user;
        password = TEST_ENV.systemTestProperties.db2.password;
        database = TEST_ENV.systemTestProperties.db2.database;
    });

    afterAll(async () => {
        await TestEnvironment.cleanUp(TEST_ENV);
    });


    it("should export the system roles table to the standard out without profile just using cmd options", () => {
        const regex: RegExp = new RegExp(fs.readFileSync(__dirname + "/__regex__/export_table.regex").toString(), "gs");
        const response = runCliScript(__dirname + "/__scripts__/success_no_profile.sh", TEST_ENV,
            [host, port, user, password, database]);
        expect(response.stderr.toString()).toBe("");
        expect(response.status).toBe(0);
        expect(new RegExp(regex, "gs").test(response.stdout.toString())).toBe(true);
    });
});
