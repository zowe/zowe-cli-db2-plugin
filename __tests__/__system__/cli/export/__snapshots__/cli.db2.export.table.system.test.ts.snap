// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`db2 export table command should display the help 1`] = `
"
 COMMAND NAME
 ------------

   table

 DESCRIPTION
 -----------

   Export a Db2 table to the stdout or a file.

 USAGE
 -----

   zowe db2 export table <table> [options]

 POSITIONAL ARGUMENTS
 --------------------

   table		 (string)

      The name of the table to export

 OPTIONS
 -------

   --outfile  | -o (string)

      The path to the output file

   --separator  | --sep (string)

      Specify whether to add a separator between statements when exporting a table

 DB2 CONNECTION OPTIONS
 ----------------------

   --host  | -H (string)

      The Db2 server host name

   --port  | -P (number)

      The Db2 server port number

   --user  | -u (string)

      The Db2 user ID (may be the same as the TSO login)

   --password  | --pass | --pw (string)

      The Db2 password (may be the same as the TSO password)

   --database  | --db (string)

      The name of the database

   --sslFile  | --ssl (string)

      Path to an SSL Certificate file

 PROFILE OPTIONS
 ---------------

   --db2-profile  | --db2-p (string)

      The name of a (db2) profile to load for this command execution.

   --base-profile  | --base-p (string)

      The name of a (base) profile to load for this command execution.

 BASE CONNECTION OPTIONS
 -----------------------

   --reject-unauthorized  | --ru (boolean)

      Reject self-signed certificates.

      Default value: true

   --token-type  | --tt (string)

      The type of token to get and use for the API. Omit this option to use the
      default token type, which is provided by 'zowe auth login'.

   --token-value  | --tv (string)

      The value of the token to pass to the API.

 GLOBAL OPTIONS
 --------------

   --response-format-json  | --rfj (boolean)

      Produce JSON formatted data from a command

   --help  | -h (boolean)

      Display help text

   --help-examples  (boolean)

      Display examples for all the commands in a the group

   --help-web  | --hw (boolean)

      Display HTML help in browser

 EXAMPLES
 --------

   - Export employees data from the table SAMPLE.EMP and save it
   to the file 'employees.sql':

      $ zowe db2 export table SAMPLE.EMP --outfile employees.sql

{
  \\"success\\": true,
  \\"exitCode\\": 0,
  \\"message\\": \\"The help was constructed for command: table.\\",
  \\"stdout\\": \\"\\\\n COMMAND NAME\\\\n ------------\\\\n\\\\n   table\\\\n\\\\n DESCRIPTION\\\\n -----------\\\\n\\\\n   Export a Db2 table to the stdout or a file.\\\\n\\\\n USAGE\\\\n -----\\\\n\\\\n   zowe db2 export table <table> [options]\\\\n\\\\n POSITIONAL ARGUMENTS\\\\n --------------------\\\\n\\\\n   table\\\\t\\\\t (string)\\\\n\\\\n      The name of the table to export\\\\n\\\\n OPTIONS\\\\n -------\\\\n\\\\n   --outfile  | -o (string)\\\\n\\\\n      The path to the output file\\\\n\\\\n   --separator  | --sep (string)\\\\n\\\\n      Specify whether to add a separator between statements when exporting a table\\\\n\\\\n DB2 CONNECTION OPTIONS\\\\n ----------------------\\\\n\\\\n   --host  | -H (string)\\\\n\\\\n      The Db2 server host name\\\\n\\\\n   --port  | -P (number)\\\\n\\\\n      The Db2 server port number\\\\n\\\\n   --user  | -u (string)\\\\n\\\\n      The Db2 user ID (may be the same as the TSO login)\\\\n\\\\n   --password  | --pass | --pw (string)\\\\n\\\\n      The Db2 password (may be the same as the TSO password)\\\\n\\\\n   --database  | --db (string)\\\\n\\\\n      The name of the database\\\\n\\\\n   --sslFile  | --ssl (string)\\\\n\\\\n      Path to an SSL Certificate file\\\\n\\\\n PROFILE OPTIONS\\\\n ---------------\\\\n\\\\n   --db2-profile  | --db2-p (string)\\\\n\\\\n      The name of a (db2) profile to load for this command execution.\\\\n\\\\n   --base-profile  | --base-p (string)\\\\n\\\\n      The name of a (base) profile to load for this command execution.\\\\n\\\\n BASE CONNECTION OPTIONS\\\\n -----------------------\\\\n\\\\n   --reject-unauthorized  | --ru (boolean)\\\\n\\\\n      Reject self-signed certificates.\\\\n\\\\n      Default value: true\\\\n\\\\n   --token-type  | --tt (string)\\\\n\\\\n      The type of token to get and use for the API. Omit this option to use the\\\\n      default token type, which is provided by 'zowe auth login'.\\\\n\\\\n   --token-value  | --tv (string)\\\\n\\\\n      The value of the token to pass to the API.\\\\n\\\\n GLOBAL OPTIONS\\\\n --------------\\\\n\\\\n   --response-format-json  | --rfj (boolean)\\\\n\\\\n      Produce JSON formatted data from a command\\\\n\\\\n   --help  | -h (boolean)\\\\n\\\\n      Display help text\\\\n\\\\n   --help-examples  (boolean)\\\\n\\\\n      Display examples for all the commands in a the group\\\\n\\\\n   --help-web  | --hw (boolean)\\\\n\\\\n      Display HTML help in browser\\\\n\\\\n EXAMPLES\\\\n --------\\\\n\\\\n   - Export employees data from the table SAMPLE.EMP and save it\\\\n   to the file 'employees.sql':\\\\n\\\\n      $ zowe db2 export table SAMPLE.EMP --outfile employees.sql\\\\n\\\\n\\",
  \\"stderr\\": \\"\\",
  \\"data\\": \\"\\\\n COMMAND NAME\\\\n ------------\\\\n\\\\n   table\\\\n\\\\n DESCRIPTION\\\\n -----------\\\\n\\\\n   Export a Db2 table to the stdout or a file.\\\\n\\\\n USAGE\\\\n -----\\\\n\\\\n   zowe db2 export table <table> [options]\\\\n\\\\n POSITIONAL ARGUMENTS\\\\n --------------------\\\\n\\\\n   table\\\\t\\\\t (string)\\\\n\\\\n      The name of the table to export\\\\n\\\\n OPTIONS\\\\n -------\\\\n\\\\n   --outfile  | -o (string)\\\\n\\\\n      The path to the output file\\\\n\\\\n   --separator  | --sep (string)\\\\n\\\\n      Specify whether to add a separator between statements when exporting a table\\\\n\\\\n DB2 CONNECTION OPTIONS\\\\n ----------------------\\\\n\\\\n   --host  | -H (string)\\\\n\\\\n      The Db2 server host name\\\\n\\\\n   --port  | -P (number)\\\\n\\\\n      The Db2 server port number\\\\n\\\\n   --user  | -u (string)\\\\n\\\\n      The Db2 user ID (may be the same as the TSO login)\\\\n\\\\n   --password  | --pass | --pw (string)\\\\n\\\\n      The Db2 password (may be the same as the TSO password)\\\\n\\\\n   --database  | --db (string)\\\\n\\\\n      The name of the database\\\\n\\\\n   --sslFile  | --ssl (string)\\\\n\\\\n      Path to an SSL Certificate file\\\\n\\\\n PROFILE OPTIONS\\\\n ---------------\\\\n\\\\n   --db2-profile  | --db2-p (string)\\\\n\\\\n      The name of a (db2) profile to load for this command execution.\\\\n\\\\n   --base-profile  | --base-p (string)\\\\n\\\\n      The name of a (base) profile to load for this command execution.\\\\n\\\\n BASE CONNECTION OPTIONS\\\\n -----------------------\\\\n\\\\n   --reject-unauthorized  | --ru (boolean)\\\\n\\\\n      Reject self-signed certificates.\\\\n\\\\n      Default value: true\\\\n\\\\n   --token-type  | --tt (string)\\\\n\\\\n      The type of token to get and use for the API. Omit this option to use the\\\\n      default token type, which is provided by 'zowe auth login'.\\\\n\\\\n   --token-value  | --tv (string)\\\\n\\\\n      The value of the token to pass to the API.\\\\n\\\\n GLOBAL OPTIONS\\\\n --------------\\\\n\\\\n   --response-format-json  | --rfj (boolean)\\\\n\\\\n      Produce JSON formatted data from a command\\\\n\\\\n   --help  | -h (boolean)\\\\n\\\\n      Display help text\\\\n\\\\n   --help-examples  (boolean)\\\\n\\\\n      Display examples for all the commands in a the group\\\\n\\\\n   --help-web  | --hw (boolean)\\\\n\\\\n      Display HTML help in browser\\\\n\\\\n EXAMPLES\\\\n --------\\\\n\\\\n   - Export employees data from the table SAMPLE.EMP and save it\\\\n   to the file 'employees.sql':\\\\n\\\\n      $ zowe db2 export table SAMPLE.EMP --outfile employees.sql\\\\n\\\\n\\"
}"
`;

exports[`db2 export table command should fail with empty table name 1`] = `
"
Syntax Error:
Missing Positional Argument: table
Argument Description: The name of the table to export

Example:

 - Export employees data from the table SAMPLE.EMP and save it to the file
 'employees.sql':

      $ zowe db2 export table SAMPLE.EMP --outfile employees.sql

Use \\"zowe db2 export table --help\\" to view command description, usage, and options.
"
`;

exports[`db2 export table command should fail with invalid option 1`] = `
"Command Error:
Unknown argument: format
Command failed due to improper syntax
Command entered: \\"zowe db2 export table SYSIBM.SYSDUMMY1 --format json\\"
Available commands are \\"table\\".
Use \\"zowe db2 export --help\\" to view groups, commands, and options.
Error Details:
Unknown argument: format
"
`;
